qclib/reversed_gates.qc

```diff
@@ -0,0 +1,13 @@
+
+
+; rev_CNOT implements a CNOT gate where the
+; the control qbit `a` have a higher qbit number
+; Question: This rather feels like a limitation
+; in the driver, maybe we should build a generic
+; reversibility there instead?
```

Comment:
I think it is good to put as few requirements on the driver as possible. That makes it easier to create a more interesting driver than what we have now. A third path is to start working on a preprocessing language that support loops etc, and that can also put a CNOT in the right direction as part of the preprocessing of the compilation step.

Reply:
On the other hand the physical location of the qubit feels like something belonging in the driver. It feels a bit odd that you should have to know which of qubit a and b that is higher when you write qc code. In a subroutine for example,  you won't know the position of the passed in variables. But maybe a 3rd stage as you suggest is the way to go, that optimizes qcode for the restrictions for a particular driver

Reply:
No I think you are right after all. It belongs in the driver, and it should just replace a CNOT it cannot handle with equivalent circuitry to accomplish the gate.